strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+    """
+    return (num_gauss ** num_modes, 2 * num_modes, 2 * num_modes)
+
+
+def to_xp(num_modes):
+    r"""Provides array of indices to order quadratures as all x
+    followed by all p, starting from (x1,p1,...,xn,pn) ordering.
+
+    Args:
+        num_modes (int): number of modes
+
+    Returns:
+        array: quadrature ordering for all x followed by all p
+    """
+    return np.concatenate((np.arange(0, 2 * num_modes, 2), np.arange(0, 2 * num_modes, 2) + 1))
+
+
+def from_xp(num_modes):
+    r"""Provides array of indices to order quadratures as (x1,p1,...,xn,pn)
+    starting from all x followed by all p.
+
+    Args:
+        num_modes (int): number of modes
+
+    Returns:
+        list: quadrature ordering for (x1,p1,...,xn,pn)
+    """
+    perm_inds_list = [(i, i + num_modes) for i in range(num_modes)]
+    perm_inds = [a for tup in perm_inds_list for a in tup]
+    return perm_inds
+
+
+def update_means(means, X, perm_out):
+    r"""Apply a linear transformation ``X`` to the array of means. The
+    quadrature ordering can be specified by ``perm_out`` to match ordering
+    of ``X`` to means.
+
+    Args:
+        means (array): array of mean vectors
+        X (array): matrix for linear transformation
+        perm_out (array): indices for quadrature ordering
+
+    Returns:
+        array: transformed array of mean vectors
+    """
+    X_perm = X[:, perm_out][perm_out, :]
+    return (X_perm @ means.T).T
+
+
+def update_covs(covs, X, perm_out, Y=0):
+    r"""Apply a linear transformation parametrized by ``(X,Y)`` to the
+    array of covariance matrices. The  quadrature ordering can be specified
+    by ``perm_out`` to match ordering of ``(X,Y)`` to the covariance matrices.
+
+    If Y is not specified, it defaults to 0.
+
+    Args:
+        covs (array): array of covariance matrices
+        X (array): matrix for mutltiplicative part of transformation
+        perm_out (array): indices for quadrature ordering
+        Y (array or 0): matrix for additive part of transformation.
+
+    Returns:
+        array: transformed array of covariance matrices
+    """
+    X_perm = X[:, perm_out][perm_out, :]
+    if not isinstance(Y, int):
+        Y = Y[:, perm_out][perm_out, :]
+    return (X_perm @ covs @ X_perm.T) + Y
+
+
+# pylint: disable=too-many-instance-attributes
+class BosonicModes:
+    """A Bosonic circuit class."""
+
+    # pylint: disable=too-many-public-methods
+
+    def __init__(self):
+        self.hbar = 2
+
+    def add_mode(self, peak_list=None):
+        r"""Add len(peak_list) modes to the circuit. Each mode has a number of
+        weights specified by peak_list, and the means and covariances are set
+        to vacuum.
+
+        Args:
+            peak_list (list): list of weights per mode.
+        """
+        if peak_list is None:
+            peak_list = [1]
+
+        num_modes = len(peak_list)
+        num_gauss = np.prod(peak_list)
+        self.nlen += num_modes
```

Comment:
How about initializing this attribute in `__init__`? (Perhaps by calling `reset`)

Reply:
I have fixed this now.